executor/aggfuncs/func_count.go

```diff
@@ -0,0 +1,110 @@
+package aggfuncs
+
+import (
+	"github.com/juju/errors"
+
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type baseCount struct {
+	baseAggFunc
+}
+
+type partialResult4Count = int64
+
+func (e *baseCount) AllocPartialResult() PartialResult {
+	return PartialResult(new(partialResult4Count))
+}
+
+func (e *baseCount) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4Count)(pr)
+	*p = 0
+}
+
+func (e *baseCount) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4Count)(pr)
+	chk.AppendInt64(e.ordinal, *p)
+	return nil
+}
+
+type countOriginal struct {
+	baseCount
+}
+
+func (e *countOriginal) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4Count)(pr)
+
+	for _, row := range rowsInGroup {
+		isNull, err := e.args[0].IsNull(sctx, row)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if isNull {
+			continue
+		}
+
+		*p++
+	}
+
+	return nil
+}
+
+type countOriginalWithDistinct struct {
+	baseCount
+
+	dc *distinctChecker
+}
+
+func (e *countOriginalWithDistinct) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) (err error) {
+	p := (*partialResult4Count)(pr)
+
+	for _, row := range rowsInGroup {
+		isNull := false
+		for _, arg := range e.args {
+			isNull, err = arg.IsNull(sctx, row)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if isNull {
+				break
+			}
+		}
+		if isNull {
+			continue
+		}
+
+		buf, err := encodeRow(sctx, e.args, row)
```

Comment:
`count(distinct)` is special, we can do it this way for convenience, although it sacrifices some performance:
 1. extract a function to evaluate an argument and encode it into a provided `[]byte`,  as for these `appendXXX` functions, for example `appendInt64`, you can refer to https://github.com/pingcap/tidb/blob/master/util/chunk/column.go#L105, which is much faster than `EncodeXXX` functions, and thus we can remove the implementation of `distinctChecker`, which is inefficient
```go
func (e *countOriginalWithDistinct) evalAndEncode(arg expression.Expression, encodedBytes []byte) ([]byte, bool, error) {
    switch e.args[i].GetType().EvalType() {
    case types.ETInt:
        val, isNull, err := e.args[i].EvalInt(ctx, row)
        if err != nil || isNull {
            return encodedBytes, isNull, errors.Trace(err)
        }
        encodedBytes = appendInt64(encodedBytes, val)
    case types.ETReal:
            ...
    }
    return encodedBytes, false, nil
}
```


 2. evaluate each argument and encode them into a `[]byte`, use that `[]byte` to check where the input parameters are duplicated:
```go
encodedBytes := make([]byte, 0, 8)
for _, row := range rowsInGroup {
    for i := 0; i < len(e.args) && !hasNull; i++  {
        encodedBytes, isNull, err = e. evalAndEncode(e.args[i], encodedBytes)
        if err != nil { return ... }
        if isNull { hasNull = true; break... }
    }
    if hasNull || p.exists[encodedBytes] {
        continue
    }
    p.exists[encodedBytes] = true
    p.count++
}
```